go-controller/pkg/ovn/controller/apbroute/network_client.go

```diff
@@ -269,50 +239,47 @@ func (nb *northBoundClient) addGWRoutesForPod(gateways []*gateway_info.GatewayIn
 	}
 
 	port := portPrefix + types.GWRouterToExtSwitchPrefix + gr
-	routeInfo, err := nb.ensureRouteInfoLocked(podNsName)
-	if err != nil {
-		return fmt.Errorf("failed to ensure routeInfo for %s, error: %v", podNsName, err)
-	}
-	defer routeInfo.Unlock()
-	for _, podIPNet := range podIfAddrs {
-		for _, gateway := range gateways {
-			// TODO (trozet): use the go bindings here and batch commands
-			// validate the ip and gateway belong to the same address family
-			gws, err := util.MatchAllIPStringFamily(utilnet.IsIPv6(podIPNet.IP), gateway.Gateways.UnsortedList())
-			if err != nil {
-				klog.Warningf("Address families for the pod address %s and gateway %s did not match", podIPNet.IP.String(), gateway.Gateways)
-				continue
-			}
-			podIP := podIPNet.IP.String()
-			for _, gw := range gws {
-				// if route was already programmed, skip it
-				if foundGR, ok := routeInfo.PodExternalRoutes[podIP][gw]; ok && foundGR == gr {
-					routesAdded++
+	return nb.externalGatewayRouteInfo.CreateOrLoad(podNsName, func(routeInfo *RouteInfo) error {
```

Comment:
Personal opinion, but I feel it's more readable the way it was before. This call to `CreateOrLoad` hides the fact that it is retrieving the routeInfo before calling the function that processes the new GW IPs.

I don't see a functional improvement here compared to the old code where it was first retrieving the routeInfo and defering the unlock.

But that's me, so if this kind of pattern is used in many places in the code then I guess it's business as usual :smile:

Reply:
I guess it may be not obvious, but `CreateOrLoad` says that is will either load an existing routeInfo or create a new one and pass to the given function similar to the `Delete` method. The advantage here is that you don't need to care about routeInfo locking, because the cache will that for you